### DISCUSSION TITLE
[API Proposal]: Add indexer  to record

### DISCUSSION THREAD

**Discussion by Mr0N** [dotnet/csharplang#7705] [General] [answered]
### Background and motivation

I suggest adding an indexer to the record for all fields so that the runtime can automatically generate code for it, similar to how it is done for ToString

### API Proposal

```csharp
    public record JsonSaveModel
    {
        public object this[string obj]
        {
            set {
                if (obj == nameof(index))
                    index = value;
                throw new Exception();
            }
            get {
                if (obj == nameof(index))
                    return index;
                throw new Exception();
            }
        }
      
        public string index { get; set; }
}
```

### API Usage

```csharp
 class Q
    {
        public Q()
        {
            var res = new JsonSaveModel();
            res["index"] = "90239023";
        }
    }
```

**Comment by hez2010** [comment]
This should be a language proposal instead of an API proposal.

**Comment by HaloFour** [comment]
This sounds like a good use case for a source generator.  I don't think indexers as you've described would be a generally desirable feature.

**Reply by CyrusNajmabadi** [reply]
Agreed with Halo.  This would be very simple to do with a generator :)

**Reply by Mr0N** [reply]
Due to the not very high speed of reflection, it seems to me that such a function will sometimes be useful

**Reply by CyrusNajmabadi** [reply]
A source generator does not need reflection (either to function, or for the code it generates :)).  This would just generate a switch for you that would allow you to lookup members in the type by their string name :)

**Reply by Mr0N** [reply]
I have to write a source generator for each project. It's easier to write the indexer manually

**Reply by CyrusNajmabadi** [reply, marked as answer]
@Mr0N you write the source generator once. Then you can use it in every project.

**Reply by Serdan** [reply]
@Mr0N https://github.com/Serdan/RecordIndexerGenerator

I won't maintain it, but that should get you started. I've prepared it for publication on nuget as well. Just build and upload the package.

**Reply by Mr0N** [reply]
I've almost finished the project with indexers, and the ChatGPT indexer wrote to me. I wrote it this way for the future; it might be useful to someone because reflection doesn't work in all cases. Okay, if you ever want to do it.

**Reply by Mr0N** [reply]
![image](https://github.com/dotnet/csharplang/assets/61237140/8de9fcbc-1c6c-4728-8dfa-631af21a17c0)
Something like this is happening with the indexer in the records
